### PR TITLE
`HashedRunFinder` run-finding improvements

### LINKED ISSUE #4960: Something is terribly bad with aggregation hash performance in this test case
Trove hashing:
```
io.deephaven.engine.table.impl.QueryTable.DISABLE_PARALLEL_WHERE=true
hash=new gnu.trove.set.hash.TIntHashSet()
long start = System.nanoTime(); t=emptyTable(30_000_000).view("someIntColumn=i%10_000").where("hash.add(someIntColumn)"); long end = System.nanoTime(); println ("Time=" + ((end-start)/1_000_000_000D) + " seconds")
Time=0.190421208 seconds
```

Our hashing:
```
long start = System.nanoTime(); t=emptyTable(30_000_000).view("someIntColumn=i%10_000").countBy("Count", "someIntColumn"); long end = System.nanoTime(); println ("Time=" + ((end-start)/1_000_000_000D) + " seconds")
Time=9.652281667 seconds
```

Just crashing through the data:
```
long start = System.nanoTime(); t=emptyTable(30_000_000).view("someIntColumn=i%10_000").sumBy(); long end = System.nanoTime(); println ("Time=" + ((end-start)/1_000_000_000D) + " seconds")
Time=0.169933084 seconds
```

We should understand why this particular aggregation is so bad, and try to fix it. It's counter-intuitively terrible.

### PR DESCRIPTION
**Pull request by lbooker42**
I compared a few strategies for run-finding: 

- Linear Probe Hashing (default)
- Sorted
- No Run Finding
- Quadratic Probe Hashing
- Double Hash Hashing

Here are the results:
image

### CODE REVIEW

**Comment by lbooker42** [comment]
The performance tests were executed in a Groovy repl:
```
io.deephaven.engine.table.impl.QueryTable.DISABLE_PARALLEL_WHERE=true

QueryScope.addParam("test", (table, description) -> {
    hash=new gnu.trove.set.hash.TIntHashSet()
    start = System.nanoTime();
    t1=table.where("hash.add(someIntColumn)");
    end = System.nanoTime(); println(description + "\tTrove\t" + ((end-start)/1_000_000_000D))
    start = System.nanoTime();
    t2=table.countBy("Count", "someIntColumn");
    end = System.nanoTime(); println(description + "\tDH\t" + ((end-start)/1_000_000_000D))
})

println("10")

test(emptyTable(30_000_000).view("someIntColumn=i%10"), "Ascending0")
test(emptyTable(30_000_000).view("someIntColumn=i%10"), "Ascending1")
test(emptyTable(30_000_000).view("someIntColumn=i%10"), "Ascending2")
test(emptyTable(30_000_000).view("someIntColumn=i%10"), "Ascending3")
test(emptyTable(30_000_000).view("someIntColumn=i%10"), "Ascending4")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10 + 10)"), "Descending0")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10 + 10)"), "Descending1")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10 + 10)"), "Descending2")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10 + 10)"), "Descending3")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10 + 10)"), "Descending4")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10)"), "Random0")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10)"), "Random1")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10)"), "Random2")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10)"), "Random3")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10)"), "Random4")

println("1_000")

test(emptyTable(30_000_000).view("someIntColumn=i%1_000"), "Ascending0")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000"), "Ascending1")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000"), "Ascending2")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000"), "Ascending3")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000"), "Ascending4")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000 + 1_000)"), "Descending0")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000 + 1_000)"), "Descending1")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000 + 1_000)"), "Descending2")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000 + 1_000)"), "Descending3")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000 + 1_000)"), "Descending4")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000)"), "Random0")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000)"), "Random1")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000)"), "Random2")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000)"), "Random3")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000)"), "Random4")

println("10_000")

test(emptyTable(30_000_000).view("someIntColumn=i%10_000"), "Ascending0")
test(emptyTable(30_000_000).view("someIntColumn=i%10_000"), "Ascending1")
test(emptyTable(30_000_000).view("someIntColumn=i%10_000"), "Ascending2")
test(emptyTable(30_000_000).view("someIntColumn=i%10_000"), "Ascending3")
test(emptyTable(30_000_000).view("someIntColumn=i%10_000"), "Ascending4")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10_000 + 10_000)"), "Descending0")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10_000 + 10_000)"), "Descending1")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10_000 + 10_000)"), "Descending2")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10_000 + 10_000)"), "Descending3")
test(emptyTable(30_000_000).view("someIntColumn=(-i%10_000 + 10_000)"), "Descending4")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10_000)"), "Random0")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10_000)"), "Random1")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10_000)"), "Random2")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10_000)"), "Random3")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 10_000)"), "Random4")

println("100_000")

test(emptyTable(30_000_000).view("someIntColumn=i%100_000"), "Ascending0")
test(emptyTable(30_000_000).view("someIntColumn=i%100_000"), "Ascending1")
test(emptyTable(30_000_000).view("someIntColumn=i%100_000"), "Ascending2")
test(emptyTable(30_000_000).view("someIntColumn=i%100_000"), "Ascending3")
test(emptyTable(30_000_000).view("someIntColumn=i%100_000"), "Ascending4")
test(emptyTable(30_000_000).view("someIntColumn=(-i%100_000 + 100_000)"), "Descending0")
test(emptyTable(30_000_000).view("someIntColumn=(-i%100_000 + 100_000)"), "Descending1")
test(emptyTable(30_000_000).view("someIntColumn=(-i%100_000 + 100_000)"), "Descending2")
test(emptyTable(30_000_000).view("someIntColumn=(-i%100_000 + 100_000)"), "Descending3")
test(emptyTable(30_000_000).view("someIntColumn=(-i%100_000 + 100_000)"), "Descending4")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 100_000)"), "Random0")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 100_000)"), "Random1")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 100_000)"), "Random2")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 100_000)"), "Random3")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 100_000)"), "Random4")

println("1_000_000")

test(emptyTable(30_000_000).view("someIntColumn=i%1_000_000"), "Ascending0")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000_000"), "Ascending1")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000_000"), "Ascending2")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000_000"), "Ascending3")
test(emptyTable(30_000_000).view("someIntColumn=i%1_000_000"), "Ascending4")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000_000 + 1_000_000)"), "Descending0")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000_000 + 1_000_000)"), "Descending1")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000_000 + 1_000_000)"), "Descending2")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000_000 + 1_000_000)"), "Descending3")
test(emptyTable(30_000_000).view("someIntColumn=(-i%1_000_000 + 1_000_000)"), "Descending4")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000_000)"), "Random0")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000_000)"), "Random1")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000_000)"), "Random2")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000_000)"), "Random3")
test(emptyTable(30_000_000).update("someIntColumn=randomInt(0, 1_000_000)"), "Random4")
```

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/HashedRunFinder.java [hunk 1]
```diff
@@ -72,7 +72,8 @@ public static boolean findRunsHashed(
 
         for (int chunkPosition = 0; chunkPosition < size; ++chunkPosition) {
             final int outputPosition = outputPositions.get(chunkPosition);
-            int hashSlot = outputPosition & context.tableMask;
+            // Increase the chance of hash collision, but reduce the linear probing needed in pathological cases.
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/HashedRunFinder.java:75]: Increase relative to what? This comment is more of a commit message than an educational comment in the code. Might want to include the original code in the comment, and provide more context.

**Comment by rcaudy** [comment]
I think you should run with all variations of these configuration settings and see where we're at:
```
    public static final boolean SKIP_RUN_FIND =
            Configuration.getInstance().getBooleanWithDefault("ChunkedOperatorAggregationHelper.skipRunFind", false);
    static final boolean HASHED_RUN_FIND =
            Configuration.getInstance().getBooleanWithDefault("ChunkedOperatorAggregationHelper.hashedRunFind", true);
```

**Review by cpwright — DISMISSED** [review]
Numbers indicate it makes things pretty much universally better.

Ryan's suggestion of an improved comment makes sense.

Did you consider a version that mixes the hash a bit more?  It might reduce locality and get worse in the general case.

**Comment by rcaudy** [comment]
I'm happy to merge this if @cpwright also approves and nightlies are run and passing. Separately, I want to do more benchmarking to consider whether we should disable run finding in more scenarios, but that needs to consider other aggregations and should involve working with @stanbrub .

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/HashedRunFinder.java [hunk 2]
```diff
@@ -92,8 +89,14 @@ public static boolean findRunsHashed(
                     overflowPointer += 2;
                     break;
                 } else {
-                    // linear probe
-                    hashSlot = (hashSlot + 1) & context.tableMask;
+                    if (probe == 0) {
+                        // double hashing
+                        probe = 1 + (outputPosition % (context.tableSize - 2));
```
Comment (cpwright) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/by/HashedRunFinder.java:94]: Are we sure this probe sequence can never result in an infinite loop?

**Review by cpwright — APPROVED** [review]

File: engine/table/src/main/java/io/deephaven/engine/table/impl/by/HashedRunFinder.java (hunk 2 above)
Comment (lbooker42) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/by/HashedRunFinder.java:94]: Yes, similar to Trove we force `tableSize` to be a prime number and our `h2()` function returns a number in the range `1 <= n < tableSize` which guarantees the second hash has no common factors other than 1 (i.e. relatively prime).

I was initially worried about small desired capacities (0,1,2) but have verified that `PrimeFinder#nextPrime()` returns 3 as the smallest possible prime (i.e. when provided `0 <= x <= 3`). so the small cases are covered properly as well.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**